subcommand/delete-completed-job/command.go

```diff
@@ -0,0 +1,147 @@
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
+	"os"
+	"sync"
+	"time"
+)
+
+// Command is the command for deleting completed jobs.
+type Command struct {
+	UI cli.Ui
+
+	flags         *flag.FlagSet
+	k8s           *k8sflags.K8SFlags
+	flagNamespace string
+
+	once      sync.Once
+	help      string
+	k8sClient kubernetes.Interface
+
+	// retryDuration is how often we'll retry deletion.
+	retryDuration time.Duration
+}
+
+func (c *Command) init() {
+	c.flags = flag.NewFlagSet("", flag.ContinueOnError)
+
+	c.k8s = &k8sflags.K8SFlags{}
+	c.flags.StringVar(&c.flagNamespace, "k8s-namespace", "",
+		"Name of Kubernetes namespace where the job is deployed")
+	flags.Merge(c.flags, c.k8s.Flags())
+	c.help = flags.Usage(help, c.flags)
+
+	// Default retry to 5s. This is exposed for setting in tests.
+	if c.retryDuration == 0 {
+		c.retryDuration = 5 * time.Second
+	}
+}
+
+// Run will attempt to delete the job once it succeeds. If the job hits its
+// backoff limit, it will give up deleting it.
+func (c *Command) Run(args []string) int {
+	c.once.Do(c.init)
+
+	// Validate command.
+	if err := c.flags.Parse(args); err != nil {
+		return 1
+	}
+	if len(c.flags.Args()) != 1 {
+		c.UI.Error(fmt.Sprintf("Must have one arg: the job name to delete."))
+		return 1
+	}
+	jobName := c.flags.Args()[0]
+
+	// c.k8sclient might already be set in a test.
+	if c.k8sClient == nil {
+		config, err := subcommand.K8SConfig(c.k8s.KubeConfig())
+		if err != nil {
+			c.UI.Error(fmt.Sprintf("Error retrieving Kubernetes auth: %s", err))
+			return 1
+		}
+
+		c.k8sClient, err = kubernetes.NewForConfig(config)
+		if err != nil {
+			c.UI.Error(fmt.Sprintf("Error initializing Kubernetes client: %s", err))
+			return 1
+		}
+	}
+
+	logger := hclog.New(&hclog.LoggerOptions{
+		Level:  hclog.Info,
+		Output: os.Stderr,
+	})
+
+	// Wait for job to complete.
+	logger.Info(fmt.Sprintf("waiting for job %q to complete successfully", jobName))
+	for {
+		job, err := c.k8sClient.BatchV1().Jobs(c.flagNamespace).Get(jobName, metav1.GetOptions{})
+		if k8serrors.IsNotFound(err) {
+			logger.Info(fmt.Sprintf("job %q does not exist, no need to delete", jobName))
+			return 0
+		}
+		if err != nil {
+			c.UI.Error(fmt.Sprintf("Error getting job %q: %s", jobName, err))
+			return 1
+		}
+
+		// If its succeeded we're done.
+		if job.Status.Succeeded > 0 {
+			break
+		}
+
+		// If its reached its backoff limit then it will never complete,
+		// exit and delete ourselves.
+		for _, condition := range job.Status.Conditions {
+			if condition.Type == v1.JobFailed && condition.Reason == "BackoffLimitExceeded" {
+				logger.Warn(fmt.Sprintf("job %q has reached its backoff limit and will never complete", jobName))
+				return 0
+			}
+		}
+
+		logger.Info(fmt.Sprintf("job %q has not yet succeeded, waiting %v", jobName, c.retryDuration))
+		time.Sleep(c.retryDuration)
+	}
```

Comment:
I feel like there could be cases when the ACL init job will keep running forever and not exit. There are few times I ran into this. The first time was when the ACL init job was trying to connect to the consul server on the wrong IP. The second time was when the job didn't have the right permissions to write a secret. Although we can fix these issues (I believe this PR fixes the second one), it seems like there could be other edge cases when this could happen. 

In case the ACL init job runs forever, we should prevent the delete task from running forever as well.

Reply:
Should we have a timeout then, you think? Maybe 10 minutes?

Reply:
I've added a timeout that defaults to 10m to the delete command. I also added a timeout to the acl-init command. For the latter, I used the context package. This way if it gets into a weird state it will eventually exit. It added a lot of code though.

Reply:
@lkysow I think we need to change default timeouts for both commands to be less than 5min (helm's default timeout). 

I had a deployment where servers never became healthy. The ACL init job has timed out eventually and finished with error, as expected. But the cleanup container got deleted by helm before it had a chance to delete the ACL init job. This happened because `helm upgrade` or `helm install` will wait until the hook job will finish or until a timeout is reached. The default helm timeout of 5min is smaller than the default timeouts for both commands, so helm considered the hook "failed" and deleted the job.

Reply:
Alternatively, we could change `hook-delete-policy` to `before-hook-creation`, so that helm will delete it only on the next install or upgrade.

Reply:
Another issue is that in the worst case it'll take the ACL init job to run 60min by default (the default backoff limit of 6 x job's default timeout of 10min). The default timeout for the ACL cleanup job should be more than that.